### PR TITLE
add format_list_item.data.frame

### LINKED ISSUE #6592: format_col and nested 1-column frames
(This may be related to https://github.com/Rdatatable/data.table/issues/5948, since nested frames is a common link.)

```r
mt <- as.data.table(mtcars)[1:3,]
mt$frm <- list(data.frame(a=1), data.frame(a=1), data.frame(a=1))
mt
# Error in vapply(X = x, FUN = fun, ..., FUN.VALUE = NA_character_, USE.NAMES = use.names) : 
#   values must be type 'character',
#  but FUN(X[[1]]) result is type 'list'
mt[,-12]
#      mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
#    <num> <num> <num> <num> <num> <num> <num> <num> <num> <num> <num>
# 1:  21.0     6   160   110  3.90 2.620 16.46     0     1     4     4
# 2:  21.0     6   160   110  3.90 2.875 17.02     0     1     4     4
# 3:  22.8     4   108    93  3.85 2.320 18.61     1     1     4     1
```

When debugging,

```r
debug(data.table:::format_col.default)
mt
#### 'c'ontinue through the first 11 columns, then
x
# [[1]]
#   a
# 1 1
# [[2]]
#   a
# 1 1
# [[3]]
#   a
# 1 1
vapply_1c(x, format_list_item, ...)
# Error in vapply(X = x, FUN = fun, ..., FUN.VALUE = NA_character_, USE.NAMES = use.names) : 
#   values must be type 'character',
#  but FUN(X[[1]]) result is type 'list'
format_list_item(x[[1]])
#   a
# 1 1
```

Where it is assumed that all return values from `vapply_1c` are expected to be strings.

Note that this does _not_ fail when the nested frames are more than one column, since `format_list_item.default` perhaps-naively uses `length(format(..)) == 1`.

```r
mt$frm <- list(data.frame(a=1,b=1), data.frame(), data.frame(a=1,b=1))
mt
#      mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb               frm
#    <num> <num> <num> <num> <num> <num> <num> <num> <num> <num> <num>            <list>
# 1:  21.0     6   160   110  3.90 2.620 16.46     0     1     4     4 <data.frame[1x2]>
# 2:  21.0     6   160   110  3.90 2.875 17.02     0     1     4     4 <data.frame[0x0]>
# 3:  22.8     4   108    93  3.85 2.320 18.61     1     1     4     1 <data.frame[1x2]>
```

I don't know if it makes sense to define `format_list_item.data.frame` as well to preclude this, or if there are better methods:

```r
format_list_item.data.frame <- function(x, ...) "<multi-column>"
vapply_1c(x, format_list_item.data.frame, ...)
# [1] "<multi-column>" "<multi-column>" "<multi-column>"
```

This test was done with `data.table_1.16.2`, but it also fails with `data.table_1.15.2`, so it's not a recent breakage.

```r
packageVersion("data.table")
# [1] ‘1.15.2’
mt <- as.data.table(mtcars)[1:3,]
mt$frm <- list(data.frame(a=1), data.frame(a=1), data.frame(a=1))
mt
# Error in vapply(X = x, FUN = fun, ..., FUN.VALUE = NA_character_, USE.NAMES = use.names) :
#   values must be type 'character',
#  but FUN(X[[1]]) result is type 'list'
```

<details><summary> sessionInfo() </summary>

<pre>
sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: x86_64-pc-linux-gnu (64-bit)
# Running under: Ubuntu 24.04.1 LTS
# Matrix products: default
# BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
# LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0
# locale:
#  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
#  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C           LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
# time zone: America/New_York
# tzcode source: system (glibc)
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# other attached packages:
# [1] data.table_1.16.2 r2_0.11.0        
# loaded via a namespace (and not attached):
#  [1] compiler_4.3.3    clipr_0.8.0       fastmap_1.2.0     cli_3.6.2         tools_4.3.3       htmltools_0.5.8.1 rmarkdown_2.26    knitr_1.45       
#  [9] xfun_0.42         digest_0.6.34     rlang_1.1.3       evaluate_0.23    
</pre>
</details>

### PR DESCRIPTION
**Pull request by r2evans**
Closes #6592, nested 1-column frames break print.data.table

### CODE REVIEW

**Comment by MichaelChirico** [comment]
Can you check if this fixes #5948, or if the fix there is related?

File: inst/tests/tests.Rraw [hunk 1]
```diff
@@ -20596,3 +20596,18 @@ test(2295.3, is.data.table(d2))
...
+test(2297.01, format_list_item(data.frame(a=1)),                     output = "<data.frame\\[1x1]>")
+test(2297.02, format_list_item(data.frame(a=1)[0,,drop=FALSE]),      output = "<data.frame\\[0x1]>")
+test(2297.03, format_list_item(data.frame(a=1)[,0]),                 output = "<data.frame\\[1x0]>")
+test(2297.04, format_list_item(data.frame(a=1, b=2)[0,,drop=FALSE]), output = "<data.frame\\[0x2]>")
+test(2297.05, format_list_item(data.frame(a=1, b=2)[,0]),            output = "<data.frame\\[1x0]>")
+test(2297.06, format_list_item(data.table(a=1)),                     output = "<data.table\\[1x1]>")
+test(2297.07, format_list_item(data.table(a=1)[0,]),                 output = "<data.table\\[0x1]>")
+test(2297.08, format_list_item(data.table(a=1)[,0]),                 output = "<data.table\\[0x0]>")
+test(2297.09, format_list_item(data.table(a=1, b=2)[0,]),            output = "<data.table\\[0x2]>")
+test(2297.10, format_list_item(data.table(a=1, b=2)[,0]),            output = "<data.table\\[0x0]>")
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/tests/tests.Rraw:20610]: I think the two `[1x0]` and `[0x0]` tests are redundant, right? removed in the suggested fix.

**Comment by MichaelChirico** [comment]
Thanks LGTM besides the small tweak to tests & request for whether this can fix a 2nd bug.

File: inst/tests/tests.Rraw (hunk 1 above)
[the remark this answers is not in the document]
Comment (r2evans) [reply inst/tests/tests.Rraw:20610]: I tried to do a simple expansion of 0, 1, and >1 for number of columns, and 0, 1 rows, not knowing if you were aware of other corner-cases. (This is partly because I discovered during testing that `data.frame(b=1)[,0]` is `1x0` whereas `data.table(b=1)[,0]` is `0x0`. Not sure if that's intentional.) I'm happy to remove them.

**Comment by r2evans** [comment]
This does not affect #5948, I think my comment there covers it. LMK if there's something else for this PR.

File: inst/tests/tests.Rraw (hunk 1 above)
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply inst/tests/tests.Rraw]: Yes, it's intentional -- #2422 :)

**Comment by r2evans** [comment]
stby, I didn't QA the changes, I'll fix that fail

**Review by MichaelChirico — APPROVED** [review]

**Comment by MichaelChirico** [comment]
Thanks @r2evans! I've sent you an invite to join as a project member. Most notably, that will let you create branches directly on this repo going forward which simplifies collaboration a bit. I'm also adding you as a contributor to the DESCRIPTION
